R/extract_drug_dispenses.R

```diff
@@ -195,70 +217,93 @@ extract_drug_dispenses <- function(
-        sql(soi_dtd_condition)
+    query <- er_prs_f |>
+      dplyr::inner_join(er_pha_f, by = dcir_join_keys) |>
+      dplyr::inner_join(ir_pha_r, by = c("PHA_PRS_C13" = "PHA_CIP_C13")) |>
+      dplyr::filter(
+        dbplyr::sql(soi_dtd_condition)
       )
 
+    # Dans les tables archivées par années, toutes les délivrances dont les
+    # dates EXE_SOI_DTD appartiennent à l'année d'archivage en question sont
+    # présentes. Il n'est donc pas nécessaire de filtrer sur les dates de
+    # remontée FLX_DIS_DTD pour les années archivées, sauf si la dernière année
+    # à extraire est atteinte, auquel cas il faut prendre en compte le seuil
+    # pour les dates de remontée FLX_DIS_DTD défini par dis_dtd_end_date.
+    # A l'inverse, pour la table  ER_PRS_F non archivée, un découpage selon
+    # les dates de remontée FLX_DIS_DTD est mis en place pour faire une
+    # extraction année par année.
+    # NB : Ce choix induit une quasi exhaustivité des données pour les années
+    # archivées, par opposition aux années très récentes pour lesquelles les
+    # données ne sont pas toutes remontées. Il pourrait être souhaitable
+    # d'introduire un filtre pour être à flux constant sur l'ensemble des années
+    if ((year >= first_non_archived_year) || (year == end_year)) {
+      query <- query |>
+        dplyr::filter(
+          dbplyr::sql(dis_dtd_condition)
+        )
+    }
+
     if (!is.null(atc_cod_starts_with)) {
-      query <- query %>%
-        filter(sql(atc_conditions))
+      query <- query |>
+        dplyr::filter(dbplyr::sql(atc_conditions))
     }
 
     cols_to_select <- c(
       "EXE_SOI_DTD",
       "PHA_ACT_QSN",
       "PHA_ATC_CLA",
-      "PHA_CIP_C13",
+      "PHA_PRS_C13",
       "PSP_SPE_COD"
     )
 
-    query <- query %>%
-      mutate(PHA_CIP_C13 = PHA_PRS_C13) %>%
-      select(
+    query <- query |>
+      dplyr::select(
         BEN_NIR_PSA,
-        all_of(cols_to_select)
-      ) %>%
-      distinct()
+        dplyr::all_of(cols_to_select)
+      ) |>
+      dplyr::distinct()
 
     if (!is.null(patients_ids)) {
       patients_ids_table <- dplyr::tbl(conn, patients_ids_table_name)
-      patients_ids_table <- patients_ids_table %>%
-        select(BEN_IDT_ANO, BEN_NIR_PSA) %>%
-        distinct()
-      query <- query %>%
-        inner_join(patients_ids_table, by = "BEN_NIR_PSA") %>%
-        select(
+      patients_ids_table <- patients_ids_table |>
+        dplyr::select(BEN_IDT_ANO, BEN_NIR_PSA) |>
+        dplyr::distinct()
+      query <- query |>
+        dplyr::inner_join(patients_ids_table, by = "BEN_NIR_PSA") |>
+        dplyr::select(
           BEN_IDT_ANO,
-          all_of(cols_to_select)
-        ) %>%
-        distinct()
+          dplyr::all_of(cols_to_select)
+        ) |>
+        dplyr::distinct()
     }
 
     create_table_or_insert_from_query(
       conn = conn,
-      output_table_name = table_name,
+      output_table_name = output_table_name,
       query = query
     )
```

Comment:
Je crois qu'il faudrait qu'on se repenche sur la gestion des `append = TRUE`/`FALSE`, `overwrite = TRUE`/`FALSE`. Je note dans #32 pour ce pas bloquer cette PR.

Reply:
Il ya a une incompatibilité avec la nouvelle version de `create_table_or_insert_from_query` implémentée par @strayMat : 
Je ne sais pas si cette fonction utilitaire est très pertinente mais à la base elle était sensée faire un CREATE TABLE si le nom de la table de sortie n'existe pas et INSERT INTO si elle existe. C'était justement pour être utilisé dans une boucle comme celle-ci. Rajouter un paramètre `append` à `create_table_or_insert_from_query` empêche de l'utiliser dans une boucle sur les tables à extraire sans distinguer le cas de la première itération des itérations suivantes

Reply:
A la réflexion, le regroupement de CREATE TABLE  et de INSERT INTO au sein d'une même fonction me parait apporter plus en confusion qu'en concision. Je propose les modif suivante :
- Rajout d'une option `overwrite` à `create_table_from_query`
- Suppression de `create_table_or_insert_from_query` et remplacement par une fonction `insert_into_table_from_query`
@ThomasSoeiro @strayMat ça serait ok pour vous ?

Reply:
Je m'étais dit que ça coincé mais j'ai pas pris le temps d'y réfléchir plus...

Je pense aussi que la fonction actuelle va plus nous embêter qu'autre chose et je ne crois pas que faire des fonctions soit utile ici. Pourquoi pas simplement `DBI::dbExecute(conn, glue::glue("INSERT INTO {output_table_name} {query}"), overwrite = overwrite)` et `DBI::dbExecute(conn, glue::glue("CREATE TABLE {output_table_name} AS {query}"), append = append)` selon le cas ? (Je ne sais pas s'il y a besoin de `overwrite` et `append`.)

Reply:
J'ouvre un nouveau ticket pour pouvoir en discuter plus en détail
Je laisse ma nouvelle fonction `insert_into_table_from_query` ainsi que l'ancienne `create_table_or_insert_from_query` dans la PR et je propose pour qu'on décide ce point dans une PR ultérieure qui touchera de toute façon toutes les fonctions d'extraction